src/cppext.cpp

```diff
@@ -313,6 +313,285 @@ cdd cdd_past(const cdd& state)
+ * @param varsMatrix the matrix containing the traces with the boolean variable id numbers.
+ * @param valuesMatrix the matrix containing the traces with the boolean variable values.
+ * @param current_step the current step number (= depth of the trace from the top BDD node).
+ * @param negated Whether this node is reached by (an odd number of) negated node(s).
+ * @param num_bools the number of boolean variables.
+ */
+void cdd_bdd_to_matrix_rec(ddNode* r, dynamic_two_dim_matrix* varsMatrix, dynamic_two_dim_matrix* valuesMatrix,
+                           int32_t current_step, bool negated)
+{
+    // The four terminating cases.
+    if (r == cddtrue && !negated) {
+        // Nothing special has to be done.
+        return;
+    }
+    if (r == cddtrue && negated) {
+        // We cannot delete the row here already, as we might still copy part of the
+        // row (trace) in the depth-first search. Therefore, we only mark the current row
+        // as to be ignored.
+        varsMatrix->ignore_current_row();
+        valuesMatrix->ignore_current_row();
+        return;
+    }
+    if (r == cddfalse && negated) {
+        // Nothing special has to be done.
+        return;
+    }
+    if (r == cddfalse && !negated) {
+        // We cannot delete the row here already, as we might still copy part of the
+        // row (trace) in the depth-first search. Therefore, we only mark the current row
+        // as to be ignored.
+        varsMatrix->ignore_current_row();
+        valuesMatrix->ignore_current_row();
+        return;
+    }
+
+    // We have not reached the end of a trace.
+    assert(!cdd_isterminal(r));
+    if (cdd_info(r)->type == TYPE_BDD) {
+        bddNode* node = bdd_node(r);
+
+        // First follow the true child of the BDD node.
+        varsMatrix->add_value_to_row(node->level, current_step);
+        valuesMatrix->add_value_to_row(1, current_step);
+        cdd_bdd_to_matrix_rec(node->high, varsMatrix, valuesMatrix, current_step + 1, negated ^ cdd_is_negated(r));
+
+        // Now follow the false child of the BDD node.
+        varsMatrix->next_row(current_step);
+        valuesMatrix->next_row(current_step);
+        valuesMatrix->add_value_to_row(0, current_step);
+        cdd_bdd_to_matrix_rec(node->low, varsMatrix, valuesMatrix, current_step + 1, negated ^ cdd_is_negated(r));
+    } else {
+        printf("not called with a BDD node");
+    }
+}
+
+/**
+ * Transform a BDD into an array representation.
+ *
+ * @param state a cdd containing only BDD or terminal nodes
+ * @return an array representation of the BDD.
+ */
+bdd_arrays cdd_bdd_to_array(const cdd& state)
+{
+    auto init_num_of_traces = std::min(1u << cdd_varnum, 100u);  // 1u<<cdd_varnum <==> 2 ^ cdd_varnum
+    int32_t initial_value = -1;
+
+    auto* varsMatrix = new dynamic_two_dim_matrix(init_num_of_traces, cdd_varnum, initial_value);
+    auto* valuesMatrix = new dynamic_two_dim_matrix(init_num_of_traces, cdd_varnum, initial_value);
```

Comment:
Consider using stack memory instead. 
If these object do not fit onto stack, then use `std::unique_ptr`.